### ISSUE TITLE
Patterns for Asynchronous Operations with async  JavaScript Jobs by CareersJS

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#5566]
Patterns for Asynchronous Operations with async &ndash; JavaScript Jobs by CareersJS<br>
<br>
https://ift.tt/2V61DOT<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<div>
<p>Last Updated: <time>Sep 19, 2019</time></p>
<figure><img src="https://careersjs.imgix.net/uploads/397361089_98267b8639_z.jpg?ixlib=jekyll-1.1.0&amp;w=300&amp;h=600&amp;fit=clip&amp;auto=format%2Cenhance&amp;s=d3f6a59e85f67354a64d5e4d17127f14">
<figcaption>Image is &ldquo;bike race!&rdquo; by ((brian)) via <a href="https://flic.kr/p/B7zBg">Flickr</a>; CC-BY-2.0</figcaption>
</figure>
<p>I remember the good old days when running asynchronous operations required you to use callbacks in an ever-deepening pit of &ldquo;callback hell&rdquo;. Those days aren&rsquo;t completely gone, but it&rsquo;s simple enough to wrap functions that use callbacks to make them use promises instead.</p>
<p>Oh yeah, promises. I remember the <em>better</em> good old days when promises became mainstream. I had worked with jQuery&rsquo;s <code>Deferred</code> objects for quite a while before promises were common enough that I was allowed to pull a promise library into our projects at work, and at that point we had Babel, so I didn&rsquo;t even need a promise library.</p>
<p>In any case, promises did as they promised (pun most definitely intended) for the most part and made asynchronous programming much more manageable. If you&rsquo;re not really up to par with your promise skills, you can read <a href="https://www.telerik.com/blogs/what-is-the-point-of-promises">more about promises here</a>. Of course, they had their weaknesses. Many times you either needed to nest your promises or send variables into an outer scope because some of the data you needed was only available within a promise handler. For example:</p>
<div>
<div>
<pre><code><span>function</span> <span>getVals</span> <span>()</span> <span>{</span>
    <span>return</span> <span>doSomethingAsync</span><span>().</span><span>then</span><span>(</span><span>function</span> <span>(</span><span>val</span><span>)</span> <span>{</span>
        <span>return</span> <span>doAnotherAsync</span><span>(</span><span>val</span><span>).</span><span>then</span><span>(</span><span>function</span> <span>(</span><span>anotherVal</span><span>)</span> <span>{</span>
            <span>// Here we need both val and anotherVal so we nested</span>
            <span>return</span> <span>val</span> <span>+</span> <span>anotherVal</span>
        <span>})</span>
    <span>})</span>
<span>}</span>

<span>// alernatively...</span>

<span>function</span> <span>getVals</span> <span>()</span> <span>{</span>
    <span>let</span> <span>value</span>

    <span>return</span> <span>doSomethingAsync</span><span>().</span><span>then</span><span>(</span><span>function</span> <span>(</span><span>val</span><span>)</span> <span>{</span>
        <span>// send val to the outer scope so others can use it</span>
        <span>value</span> <span>=</span> <span>val</span>
        <span>return</span> <span>doAnotherAsync</span><span>(</span><span>val</span><span>)</span>
    <span>}).</span><span>then</span><span>(</span><span>function</span> <span>(</span><span>anotherVal</span><span>)</span> <span>{</span>
        <span>// Here we grab value from outside</span>
        <span>return</span> <span>value</span> <span>+</span> <span>anotherVal</span>
    <span>})</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>Both of those examples are messy in their own way. Of course, they could be cleaned up a bit with arrow functions&hellip;</p>
<div>
<div>
<pre><code><span>function</span> <span>getVals</span> <span>()</span> <span>{</span>
    <span>return</span> <span>doSomethingAsync</span><span>().</span><span>then</span><span>(</span><span>val</span> <span>=&gt;</span> <span>doAnotherAsync</span><span>(</span><span>val</span><span>).</span><span>then</span><span>(</span><span>anotherVal</span> <span>=&gt;</span> <span>val</span> <span>+</span> <span>anotherVal</span><span>))</span>
<span>}</span>

<span>// alernatively...</span>

<span>function</span> <span>getVals</span> <span>()</span> <span>{</span>
    <span>let</span> <span>value</span>

    <span>return</span> <span>doSomethingAsync</span><span>()</span>
    <span>.</span><span>then</span><span>(</span><span>val</span> <span>=&gt;</span> <span>(</span><span>value</span> <span>=</span> <span>val</span><span>,</span> <span>doAnotherAsync</span><span>(</span><span>val</span><span>)))</span>
    <span>.</span><span>then</span><span>(</span><span>anotherVal</span> <span>=&gt;</span> <span>value</span> <span>+</span> <span>anotherVal</span><span>)</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>That may clean up a bit and remove some of the syntactic cruft, but that doesn&rsquo;t really offer any more readability. Thankfully, now we have gotten past those good old days and now we have <code>async</code> and <code>await</code> and we can avoid all that nonsense.</p>
<div>
<div>
<pre><code><span>async</span> <span>function</span> <span>getVals</span> <span>()</span> <span>{</span>
    <span>let</span> <span>val</span> <span>=</span> <span>await</span> <span>doSomethingAsync</span><span>()</span>
    <span>let</span> <span>anotherVal</span> <span>=</span> <span>await</span> <span>doAnotherAsync</span><span>(</span><span>val</span><span>)</span>

    <span>return</span> <span>val</span> <span>+</span> <span>anotherVal</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>That just looks so much simpler and easy to follow. I am going to assume that you&rsquo;re well aware of the existence of <code>async</code> and <code>await</code> so I won&rsquo;t patronize you much with the details of how they work, but if you need a refresher you can read more about <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function"><code>asyc/await</code> on MDN</a>. Here, we&rsquo;ll be focusing on the patterns we&rsquo;ve used in the past with Promises and how those patterns translate to <code>async/await</code>.</p>
<h2>&ldquo;Random&rdquo; Sequential Asynchronous Operations</h2>
<p>In the previous code snippets we technically already went over a pattern - &ldquo;Random&rdquo; Sequential Operations - which is what we will be talking about first. When I say random, I don&rsquo;t really mean random. What I&rsquo;m referring to is multiple functions that may or may not be related to each other, but are called separately. Put in another way, random operations aren&rsquo;t the same operation(s) performed across an entire list/array of inputs. If you are still confused, you will see what I mean in the later sections when I go over the &ldquo;non-random&rdquo; operations and you will be able to see the difference.</p>
<p>Anyway, like I said, you have already been treated with an example of our first pattern. The operations are run sequentially, meaning the second operation waits until the first one is done before starting up. Of course, this pattern can be appear different that the example above when using promises assuming we don&rsquo;t run into the situation that we did above where we need to pass multiple values forward to later operations:</p>
<div>
<div>
<pre><code><span>function</span> <span>getVals</span> <span>()</span> <span>{</span>
    <span>return</span> <span>doSomethingAsync</span><span>()</span>
    <span>.</span><span>then</span><span>(</span><span>val</span> <span>=&gt;</span> <span>doAnotherAsync</span><span>(</span><span>val</span><span>))</span>
    <span>.</span><span>then</span><span>(</span><span>anotherVal</span> <span>=&gt;</span> <span>/* We don't need 'val' here */</span> <span>2</span> <span>*</span> <span>anotherVal</span><span>)</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>No need to access <code>val</code> in that final handler, so we can just chain <code>then</code> calls and not worry about passing values to an outer scope. The cool thing, though, is that the <code>async/await</code> version of the code doesn&rsquo;t really change except that we use <code>2 *</code> instead of <code>val +</code> in the final expression:</p>
<div>
<div>
<pre><code><span>async</span> <span>function</span> <span>getVals</span> <span>()</span> <span>{</span>
    <span>let</span> <span>val</span> <span>=</span> <span>await</span> <span>doSomethingAsync</span><span>()</span>
    <span>let</span> <span>anotherVal</span> <span>=</span> <span>await</span> <span>doAnotherAsync</span><span>(</span><span>val</span><span>)</span>

    <span>return</span> <span>2</span> <span>*</span> <span>anotherVal</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>This is the type of situation where <code>async/await</code> excels: making a string of asynchronous calls act like they are synchronous. There are no little tricks, just straightforward &ldquo;do this then do that&rdquo; kind of code.</p>
<h2>&ldquo;Random&rdquo; Parallel Asynchronous Operations</h2>
<p>Alright, this time our operations are going to be run in parallel because none of the operations cares whether other operations are done yet, nor do they need a value from any other operations in order to be able to do their own work. When using promises, this is how it could be written (ignore the fact that I reused asynchronous function names but they are being used completely differently; they are just made up function names that make it obvious they are asynchronous; they are not necessarily the same functions used in earlier examples):</p>
<div>
<div>
<pre><code><span>function</span> <span>getVals</span> <span>()</span> <span>{</span>
    <span>return</span> <span>Promise</span><span>.</span><span>all</span><span>([</span><span>doSomethingAsync</span><span>(),</span> <span>doAnotherAsync</span><span>()])</span>
    <span>.</span><span>then</span><span>(</span><span>function</span> <span>([</span><span>val</span><span>,</span> <span>anotherVal</span><span>])</span> <span>{</span>
        <span>return</span> <span>val</span> <span>+</span> <span>anotherVal</span>
    <span>})</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>We use <code>Promise.all</code> because it allows us to pass in any number of promises and once they have all been resolved, it&rsquo;ll give us all the resolved values in one <code>then</code> handler. There are other options such as <code>Promise.any</code>, <code>Promise.some</code>, etc. depending on whether or not you&rsquo;re using a promise library or certain Babel plugins and of course depending on your use case and how you want to handle the output or the potential for rejected promises. In any case, the pattern is very similar, you just choose a different <code>Promise</code> method and you receive different output.</p>
<p>The sad/great thing here is that <code>async/await</code> doesn&rsquo;t allow us to move away from <code>Promise.all</code> or its constituents. It&rsquo;s sad because <code>async/await</code> hides the use of promises in the background, but then we need to explicitly use <code>Promise</code> in order to do things in parallel. It&rsquo;s great because it means we don&rsquo;t need to learn anything new; we can just keep using what we know, but remove the extra characters required to pass a callback function into <code>then</code>. Instead, we can just <code>await</code> and pretend all those parallel operations took no time at all.</p>
<div>
<div>
<pre><code><span>async</span> <span>function</span> <span>getVals</span> <span>()</span> <span>{</span>
    <span>let</span> <span>[</span><span>val</span><span>,</span> <span>anotherVal</span><span>]</span> <span>=</span> <span>await</span> <span>Promise</span><span>.</span><span>all</span><span>([</span><span>doSomethingAsync</span><span>(),</span> <span>doAnotherAsync</span><span>()])</span>
    <span>return</span> <span>val</span> <span>+</span> <span>anotherVal</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>So <code>async/await</code> is about more than removing callbacks and unnecessary nesting, etc.: it&rsquo;s about making asynchronous programming patterns look more like synchronous programming patterns so developers can wrap their minds around what the code is accomplishing more easily.</p>
<h2>Iterating Parallel Asynchronous Operations</h2>
<p>Here comes the operations that are <em>not</em> &ldquo;random&rdquo;. This is where we iterate over a set of values and perform the same asynchronous operation(s) over each value. In this parallel version, each element is being worked on at the same time. Here it is with promises:</p>
<div>
<div>
<pre><code><span>function</span> <span>doAsyncToAll</span> <span>(</span><span>values</span> <span>/* array */</span><span>)</span> <span>{</span>
    <span>return</span> <span>Promise</span><span>.</span><span>all</span><span>(</span><span>values</span><span>.</span><span>map</span><span>(</span><span>doSomethingAsync</span><span>))</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>Ok, so that&rsquo;s as simple as it gets. How do you do it with <code>async/await</code>? Actually you don&rsquo;t need to do anything! You could, but it would only make it more verbose:</p>
<div>
<div>
<pre><code><span>async</span> <span>function</span> <span>doAsyncToAll</span> <span>(</span><span>values</span> <span>/* array */</span><span>)</span> <span>{</span>
    <span>return</span> <span>await</span> <span>Promise</span><span>.</span><span>all</span><span>(</span><span>values</span><span>.</span><span>map</span><span>(</span><span>doSomethingAsync</span><span>))</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>That accomplishes absolutely nothing except adding a couple keywords that make you look like you&rsquo;re being smart and using modern JavaScript, but in actuality you&rsquo;re adding no value and the JavaScript engines will probably run this slower. If you get a little more complicated, though, <code>async/await</code> can certainly provide some benefit:</p>
<div>
<div>
<pre><code><span>function</span> <span>doAsyncToAll</span> <span>(</span><span>values</span> <span>/* array */</span><span>)</span> <span>{</span>
    <span>return</span> <span>Promise</span><span>.</span><span>all</span><span>(</span><span>values</span><span>.</span><span>map</span><span>(</span><span>val</span> <span>=&gt;</span> <span>{</span>
        <span>return</span> <span>doSomethingAsync</span><span>(</span><span>val</span><span>)</span>
        <span>.</span><span>then</span><span>(</span><span>anotherVal</span> <span>=&gt;</span> <span>doAnotherAsync</span><span>(</span><span>anotherValue</span> <span>*</span> <span>2</span><span>))</span>
    <span>}))</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>That is not terrible, but <code>async/await</code> may be better at making it clear what exactly is happening here:</p>
<div>
<div>
<pre><code><span>function</span> <span>doAsyncToAll</span> <span>(</span><span>values</span> <span>/* array */</span><span>)</span> <span>{</span>
    <span>return</span> <span>Promise</span><span>.</span><span>all</span><span>(</span><span>values</span><span>.</span><span>map</span><span>(</span><span>async</span> <span>val</span> <span>=&gt;</span> <span>{</span>
        <span>let</span> <span>anotherVal</span> <span>=</span> <span>await</span> <span>doSomethingAsync</span><span>(</span><span>val</span><span>)</span>
        <span>return</span> <span>doAnotherAsync</span><span>(</span><span>anotherValue</span> <span>*</span> <span>2</span><span>)</span>
    <span>}))</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>Personally, I believe that is clearer, at least from within the callback to <code>map</code>, but some people may be confused here. When I first started using <code>async/await</code> I saw <code>await</code> inside the callback and it made me think that these callbacks were not being fired in parallel. This is one mistake that people can often make when using <code>async/await</code> in nested function and is one instance where it is potentially less simple to understand than using promises directly. However, a little exposure can help you more easily spot when you are using nested async functions and therefore they inner function is separate from the outer function and an <code>await</code> there does not pause the outer function.</p>
<p>Moving on from there, once you add more steps to your function, you increase the complexity of reading promises and add greater usefulness for using <code>async/await</code>.</p>
<div>
<div>
<pre><code><span>function</span> <span>doAsyncToAll</span> <span>(</span><span>values</span> <span>/* array */</span><span>)</span> <span>{</span>
    <span>return</span> <span>Promise</span><span>.</span><span>all</span><span>(</span><span>values</span><span>.</span><span>map</span><span>(</span><span>val</span> <span>=&gt;</span> <span>{</span>
        <span>return</span> <span>doSomethingAsync</span><span>(</span><span>val</span><span>)</span>
        <span>.</span><span>then</span><span>(</span><span>anotherVal</span> <span>=&gt;</span> <span>doAnotherAsync</span><span>(</span><span>anotherValue</span> <span>*</span> <span>2</span><span>))</span>
    <span>}))</span>
    <span>.</span><span>then</span><span>(</span><span>newValues</span> <span>=&gt;</span> <span>newValues</span><span>.</span><span>join</span><span>(</span><span>'</span><span>,</span><span>'</span><span>))</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>Those multiple levels of <code>then</code> calls can really mess with a person&rsquo;s head, so let&rsquo;s implement this in the more modern way:</p>
<div>
<div>
<pre><code><span>async</span> <span>function</span> <span>doAsyncToAll</span> <span>(</span><span>values</span> <span>/* array */</span><span>)</span> <span>{</span>
    <span>const</span> <span>newValues</span> <span>=</span> <span>await</span> <span>Promise</span><span>.</span><span>all</span><span>(</span><span>values</span><span>.</span><span>map</span><span>(</span><span>async</span> <span>val</span> <span>=&gt;</span> <span>{</span>
        <span>let</span> <span>anotherVal</span> <span>=</span> <span>await</span> <span>doSomethingAsync</span><span>(</span><span>val</span><span>)</span>
        <span>return</span> <span>doAnotherAsync</span><span>(</span><span>anotherValue</span> <span>*</span> <span>2</span><span>)</span>
    <span>}))</span>
    <span>return</span> <span>newValues</span><span>.</span><span>join</span><span>(</span><span>'</span><span>,</span><span>'</span><span>)</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>As usual, there are ways to slim this down, but slimming down isn&rsquo;t the ultimate goal: readability and maintainability are, and that is generally where <code>async/await</code> comes in the handiest. It is often also simpler to write because we stay on the same synchronous paradigm as we are usually on.</p>
<h2>Iterating Sequential Asynchronous Operations</h2>
<p>We&rsquo;re onto our final pattern. Once again we&rsquo;re iterating over a list and applying the asynchronous operation(s) on each item in the list, but this time, we&rsquo;re only going to do them one at a time. In other words, we can&rsquo;t perform any operations on item two until we&rsquo;ve finished our operations on item 1, etc.</p>
<div>
<div>
<pre><code><span>function</span> <span>doAsyncToAllSequentially</span> <span>(</span><span>values</span><span>)</span> <span>{</span>
    <span>return</span> <span>values</span><span>.</span><span>reduce</span><span>((</span><span>previousOperation</span><span>,</span> <span>val</span><span>)</span> <span>=&gt;</span> <span>{</span>
        <span>return</span> <span>previousOperation</span><span>.</span><span>then</span><span>(()</span> <span>=&gt;</span> <span>doSomethingAsync</span><span>(</span><span>val</span><span>))</span>
    <span>},</span> <span>Promise</span><span>.</span><span>resolve</span><span>())</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>In order to accomplish the sequential order, we need to chain a <code>then</code> call off the previous operation. This could be done with <code>reduce</code>, but this is the most sane way of doing it. Note you that you need to pass in a resolved promise as the last argument to <code>reduce</code> so the first iteration has something to chain off.</p>
<p>Here, we can really see <code>async/await</code> shine again. We don&rsquo;t need to use any array methods, such as <code>reduce</code>. We just need a normal loop and to call <code>await</code> from within it:</p>
<div>
<div>
<pre><code><span>async</span> <span>function</span> <span>doAsyncToAllSequentially</span> <span>(</span><span>values</span><span>)</span> <span>{</span>
    <span>for</span> <span>(</span><span>let</span> <span>val</span> <span>of</span> <span>values</span><span>)</span> <span>{</span>
        <span>await</span> <span>doSomethingAsync</span><span>(</span><span>val</span><span>)</span>
    <span>}</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>If you&rsquo;re using <code>reduce</code> for a reason other than just to make the operations sequential, then you can continue using it. For example, if you&rsquo;re adding results of all the operations together:</p>
<div>
<div>
<pre><code><span>function</span> <span>doAsyncToAllSequentially</span> <span>(</span><span>values</span><span>)</span> <span>{</span>
    <span>return</span> <span>values</span><span>.</span><span>reduce</span><span>((</span><span>previousOperation</span><span>,</span> <span>val</span><span>)</span> <span>=&gt;</span> <span>{</span>
        <span>return</span> <span>previousOperation</span><span>.</span><span>then</span><span>(</span>
            <span>total</span> <span>=&gt;</span> <span>doSomethingAsync</span><span>(</span><span>val</span><span>).</span><span>then</span><span>(</span>
                <span>newVal</span> <span>=&gt;</span> <span>total</span> <span>+</span> <span>newVal</span>
            <span>)</span>
        <span>)</span>
    <span>},</span> <span>Promise</span><span>.</span><span>resolve</span><span>(</span><span>0</span><span>))</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>That just messes with my mind. It&rsquo;s amazing how we end up in that callback hell even with promises again. Granted, especially since we&rsquo;re using arrow functions, we could always just combine a lot of this onto one line, but that doesn&rsquo;t actually help make it easier to understand. Using <code>async/await</code> makes it pretty clear, though:</p>
<div>
<div>
<pre><code><span>async</span> <span>function</span> <span>doAsyncToAllSequentially</span> <span>(</span><span>values</span><span>)</span> <span>{</span>
    <span>let</span> <span>total</span> <span>=</span> <span>0</span>
    <span>for</span> <span>(</span><span>let</span> <span>val</span> <span>of</span> <span>values</span><span>)</span> <span>{</span>
        <span>let</span> <span>newVal</span> <span>=</span> <span>await</span> <span>doSomethingAsync</span><span>(</span><span>val</span><span>)</span>
        <span>total</span> <span>+=</span> <span>newVal</span>
    <span>}</span>
    <span>return</span> <span>total</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>And if you still like using <code>reduce</code> when pairing your arrays down to single values&hellip;</p>
<div>
<div>
<pre><code><span>async</span> <span>function</span> <span>doAsyncToAllSequentially</span> <span>(</span><span>values</span><span>)</span> <span>{</span>
    <span>return</span> <span>values</span><span>.</span><span>reduce</span><span>(</span><span>async</span> <span>(</span><span>previous</span><span>,</span> <span>val</span><span>)</span> <span>=&gt;</span> <span>{</span>
        <span>let</span> <span>total</span> <span>=</span> <span>await</span> <span>previous</span>
        <span>let</span> <span>newVal</span> <span>=</span> <span>await</span> <span>doSomethingAsync</span><span>(</span><span>val</span><span>)</span>

        <span>return</span> <span>total</span> <span>+</span> <span>newVal</span>
    <span>},</span> <span>Promise</span><span>.</span><span>resolve</span><span>(</span><span>0</span><span>))</span>
<span>}</span>
</code></pre>
</div>
</div>
<h2>Conclusion</h2>
<p>The relatively new <code>async/await</code> keywords have really changed the game when it comes to writing asynchronous code. They help us to eliminate or minimize so many of the aspects of asynchronous code writing and reading that have plagued JavaScript developers for so long: callback hell, writing in a clear and more synchronous manner, and just being able to read and write asynchronous code in a manner that is simpler to understand. Because of this, it&rsquo;s important to get a good understanding of how to use this new technology effectively, and I hope these patterns have helped you in that regard. God bless and happy coding.</p>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via careersjs.com https://careersjs.com<br>
<br>
October 26, 2019 at 11:59AM
